### PR TITLE
Dw/multi mining

### PR DESCRIPTION
**Pull request by Tresdin-commander**

### CODE REVIEW

File: switch/Cargo.toml [hunk 1]
```diff
@@ -34,3 +34,4 @@ yee-switch-rpc-servers = { package = "yee-switch-rpc-servers", path = "rpc-serve
 yee-switch-rpc = { package = "yee-switch-rpc", path = "rpc" }
 yee-runtime = { package = "yee-runtime", path = "../runtime" }
 yee-dev = { package = "yee-dev", path = "../dev" }
+yee-mining = { package = "yee-mining", path = "mining" }
```
Comment (guoxbin) [inline switch/Cargo.toml:37]: add blank line at the end of the file

File: switch/mining/src/client.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    fn send_submit_job_request(
+        &self,
+        job: &JobResult,
+        rpc: Rpc,
+    ) -> impl Future<Item=Output, Error=RpcError> {
+        let method = "mining_submitJob".to_owned();
+        let params = vec![json!(job)];
+        // debug!("submit_job params ---:{:?}", params);
```
Comment (guoxbin) [inline switch/mining/src/client.rs:132]: remove

File: switch/mining/src/client.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+    pub fn get_job_template(&self, rpc: Rpc) -> impl Future<Item=Job, Error=RpcError> {
+        let method = "mining_getJob".to_owned();
+        let params = vec![];
+        rpc.request(method, params).and_then(parse_response)
+    }
+}
+
+fn parse_response<T: serde::de::DeserializeOwned>(output: Output) -> Result<T, RpcError> {
+    //debug!("output---:{:?}", output);
```
Comment (guoxbin) [inline switch/mining/src/client.rs:164]: remove or uncomment and make log text canonical

File: switch/mining/src/gateway.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    pub shard_job_cache: Mutex<LruCache<String, JobTemplate>>,
+    pub new_work_tx: Sender<WorkMap>,
+    pub map: Config,
+}
+
+impl Gateway {
+    pub fn new(client: Client, new_work_tx: Sender<WorkMap>, map: Config) -> Gateway {
+        //init
+        let job = JobTemplate {
+            difficulty: DifficultyType::from(0x00000000) << 224,
```
Comment (guoxbin) [inline switch/mining/src/gateway.rs:59]: no need left shift

File: switch/mining/src/gateway.rs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                         2. If the RPC URL for yee miner is right.",
+                        );
+                    } else {
+                        error!("Rpc call get_job_template error: {:?}--shard num={}", err, key);
+                    }
+                }
+            }
+        }
+
+        let mut f = false; //更新标记，只要有一个分片数据更新即为true
```
Comment (guoxbin) [inline switch/mining/src/gateway.rs:121]: change variable name

Comment (guoxbin) [inline switch/mining/src/gateway.rs:121]: no Chinese

File: switch/mining/src/gateway.rs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+        if !set.is_empty() {
+            for (key, value) in set {
+                // debug!("set data---[{}] = {:?}", key, value);
+                if self.current_job_set.get(&key).unwrap().clone().raw_hash != value.raw_hash {
+                    f = true;
+                }
+                self.current_job_set.insert(key.clone(), value.clone());//最终数据全覆盖
```
Comment (guoxbin) [inline switch/mining/src/gateway.rs:129]: no Chinese

File: switch/mining/src/gateway.rs [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                    }
+                }
+            }
+        }
+
+        let mut f = false; //更新标记，只要有一个分片数据更新即为true
+
+        if !set.is_empty() {
+            for (key, value) in set {
+                // debug!("set data---[{}] = {:?}", key, value);
```
Comment (guoxbin) [inline switch/mining/src/gateway.rs:125]: remove or uncomment and make log text canonical

File: switch/mining/src/miner.rs [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        }
+    }
+
+    pub fn run(&mut self) {
+        loop {
+            self.notify_workers(WorkerMessage::Run);
+            select! {
+                recv(self.work_rx) -> msg => match msg {
+                    Ok(work) => {
+                       // println!("get new work .......");
```
Comment (guoxbin) [inline switch/mining/src/miner.rs:68]: remove

File: switch/mining/src/miner.rs [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+                        let work_id = work.work_id.clone();
+                        let merkle_root = work.merkle_root.clone();
+                        let extra_data = work.extra_data.clone();
+                       //  debug!("cache_and send_WorkerMessage: {}", work_id);
```
Comment (guoxbin) [inline switch/mining/src/miner.rs:72]: remove

File: switch/src/params.rs [hunk 10]
```diff
@@ -56,6 +56,14 @@ pub struct SwitchCommandCmd {
     #[structopt(short = "m", long = "max connections", value_name = "MAX-CONNECTIONS" ,default_value = "100")]
     pub max_connections: u16,
 
+    ///Specify miner poll interval
+    #[structopt(long = "poll-interval ", value_name = "POLL-INTERVAL" ,default_value = "6000")]
+    pub poll_interval: u64,
```
Comment (guoxbin) [inline switch/src/params.rs:61]: rename to job_refresh_interval

File: switch/src/params.rs [hunk 11]
```diff
@@ -56,6 +56,14 @@ pub struct SwitchCommandCmd {
...
+
+    /// start miner
+    #[structopt( long = "start-miner")]
+    pub start_miner: bool,
```
Comment (guoxbin) [inline switch/src/params.rs:65]: rename to "mine"

File: switch/mining/src/worker/pow.rs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+use log::{info, error, warn, debug};
+use super::Seal;
+use primitives::blake2_256;
+use parity_codec::Encode;
+
+extern crate chrono;
+
+pub const MAX_EXTRA_DATA_LENGTH: usize = 32;
+
+pub struct Dummy {
```
Comment (guoxbin) [inline switch/mining/src/worker/pow.rs:30]: change variable name Dummy

**Review by guoxbin — CHANGES_REQUESTED** [review]

File: switch/src/params.rs (hunk 11 above)
Comment (Tresdin-commander) [reply switch/src/params.rs:65]: thanks

**Pull request closed without merging**